File: 1143-longest-common-subsequence/1143-longest-common-subsequence.py
class Solution:
    def longestCommonSubsequence(self, a: str, b: str) -> int:
        n = len(a)
        m = len(b)
        
        dp = [[0]*(m+1) for i in range(n+1)]
        
        print(dp)
        for i in range(1,n+1):
            for j in range(1,m+1):
                if(a[n-i]==b[m-j]):
                    dp[i][j] = 1 + dp[i-1][j-1]
                else:
                    dp[i][j] = max(dp[i-1][j],dp[i][j-1])
        return dp[n][m]

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.cpp
class Solution {
public:
    int longestCommonSubsequence(string a, string b) {
        int n = a.size(),m = b.size();
        vector<vector<int>> dp(n+1,vector<int>(m+1));

        dp[0][0] = 0;
        for (int i = 1; i <= n; ++i)
        {
            for(int j = 1;j <= m;j++){
                if(a[n-i]==b[m-j]){
                    dp[i][j] = 1+dp[i-1][j-1];
                }
                else{
                    dp[i][j] = max(dp[i-1][j],dp[i][j-1]);
                }
            }
        }
        return dp[n][m];
    }
};